splipy/basis_eval.pyx

```diff
@@ -0,0 +1,122 @@
+from bisect import bisect_right, bisect_left
+import numpy as np
+cimport numpy as np
+import copy
+cimport cython
+from cpython cimport bool
+@cython.boundscheck(False) # turn off bounds-checking for entire function
+
+cdef inline int int_max(int a, int b): return a if a >= b else b
+cdef inline unsigned int uint_min(unsigned int a, unsigned int b): return a if a <= b else b
+cdef inline double fabs(double a):     return a if a > 0 else -a
+
+def evaluate(knots_in, order_in, eval_t_in, periodic_in, tolerance_in, derivatives_in=0, from_right=True):
+    """  Evaluate all basis functions in a given set of points.
+
+    :param knots_in:       Knot vector
+    :param order_in:       Parametrization order (polyonomial degree + 1)
+    :param eval_t_in:      The parametric coordinate(s) in which to evaluate
+    :param periodic_in:    Periodicity of basis
+    :param tolerance_in:   Knot tolerance for detecting end-point-evaluations
+    :param derivatives_in: Number of derivatives to compute
+    :param from_right:     True if evaluation should be done in the limit from above
+    :return: Two tuples of all arguments to the scipy sparse csr_matrix
+    """
+
+    # wrap everything into c-type datastructures for optimized performence
+    cdef np.ndarray knots   = knots_in
+    cdef int periodic       = periodic_in
+    cdef unsigned int p     = order_in  # knot vector order
+    cdef unsigned int n_all = len(knots) - p  # number of basis functions (without periodicity)
+    cdef unsigned int n     = len(knots) - p - (periodic+1)  # number of basis functions (with periodicity)
+    cdef unsigned int m     = len(eval_t_in)
+    cdef unsigned int d     = derivatives_in
+    cdef double start       = knots[p-1]
+    cdef double end         = knots[n_all]
+    cdef double evalT       ;
+    cdef double tol         = tolerance_in
+    cdef unsigned int mu    = 0
+    cdef np.ndarray t       = copy.deepcopy(eval_t_in)
+    cdef np.ndarray data    = np.zeros(m*p)
+    cdef np.ndarray indices = np.zeros(m*p, dtype='int32')
+    cdef np.ndarray indptr  = np.arange(0,m*p+1,p)
+    cdef np.ndarray M       = np.zeros(p)  # temp storage to keep all the function evaluations
+    cdef unsigned int k,q,j,i
+
+    if periodic >= 0:
+        # Wrap periodic evaluation into domain
+        for i in range(len(t)):
+            if t[i] < start or t[i] > end:
+                t[i] = (t[i] - start) % (end - start) + start
+    for i in range(len(t)):
+        right = from_right
+        evalT = t[i]
+        # Special-case the endpoint, so the user doesn't need to
+        if fabs(t[i] - end) < tol:
+            right = False
+        # Skip non-periodic evaluation points outside the domain
+        if t[i] < start or t[i] > end:
+            continue
+
+        # mu = index of last non-zero basis function
+        if right:
+            mu = bisect_right(knots, evalT)
```

Comment:
Be aware, you're calling into Python code here.

Reply:
But this is python-style code. How do I explicit tell the program what corresponding c++ function should be used? By the way, the answer is [binary_search](http://www.cplusplus.com/reference/algorithm/binary_search/) from `#include <algorithm>`

Reply:
I would avoid C++ and just `cdef` a bisection algorithm in the pyx-file.